### PR TITLE
Fix a bug returning 404 when display a single tag with no release

### PR DESCRIPTION
**Pull request by lunny**
Partially caused by #29149 

When use

```go
releases, err := getReleaseInfos(ctx, &repo_model.FindReleasesOptions{
		ListOptions: db.ListOptions{Page: 1, PageSize: 1},
		RepoID:      ctx.Repo.Repository.ID,
		TagNames:    []string{ctx.Params("*")},
		// only show draft releases for users who can write, read-only users shouldn't see draft releases.
		IncludeDrafts: writeAccess,
	})
```
replace
```go
release, err := repo_model.GetRelease(ctx, ctx.Repo.Repository.ID, ctx.Params("*"))
```
It missed `IncludeTags: true,`. That means this bug will be occupied only when the release is only a tag.

 - Get the right tag record when it's not a release
 - Display correct tag tab but not release tag when it's a tag.
 - The button will bring the tag name to the new page when it's a single tag page
 - the new page will automatically hide the release target inputbox when the tag name is pre filled. This should be backport to v1.21.

### CODE REVIEW

**Review by KN4CK3R — APPROVED** [review]

**Comment by KN4CK3R** [comment]
Could you have a look at the title too?
grafik
This does not look good for tags without release:
https://github.com/go-gitea/gitea/blob/850fc2516e67049ec195c72d861896b275bd09d1/templates/repo/release/list.tmpl#L20-L30

**Comment by lunny** [comment]
> Could you have a look at the title too? grafik This does not look good for tags without release:
> 
> https://github.com/go-gitea/gitea/blob/850fc2516e67049ec195c72d861896b275bd09d1/templates/repo/release/list.tmpl#L20-L30

Fixed in b7dc0944b3d5dc5296bd1fb3cdaf8468a18a7dfb

**Review by KN4CK3R — APPROVED** [review]

File: web_src/js/features/repo-release.js [hunk 1]
```diff
@@ -41,6 +42,10 @@ function initTagNameEditor() {
       showElem('#tag-target-selector');
       tagHelper.textContent = value ? newTagHelperText : defaultTagHelperText;
     }
+  };
+  hideTargetInput(tagNameInput); // update on page load because the input may have a value
+  tagNameInput.addEventListener('change', (e) => {
```
Comment (wxiaoguang) [inline web_src/js/features/repo-release.js]:
```suggestion
  tagNameInput.addEventListener('input', (e) => {
```

Use "input" event?

Comment (lunny) [reply web_src/js/features/repo-release.js]: But if user pastes content, it will not be triggered.

Comment (wxiaoguang) [reply web_src/js/features/repo-release.js]: What do you mean ? "input" not be triggered or "change" not be triggered?

File: templates/repo/release_tag_header.tmpl [hunk 2]
```diff
@@ -17,7 +17,7 @@
 			</a>
 		{{end}}
 		{{if and (not .PageIsTagList) .CanCreateRelease}}
-			<a class="ui small primary button" href="{{$.RepoLink}}/releases/new">
+			<a class="ui small primary button" href="{{$.RepoLink}}/releases/new{{if .PageIsSingleTag}}?tag={{.Title}}{{end}}">
```
Comment (wxiaoguang) [inline templates/repo/release_tag_header.tmpl]: Why not use TagName here? Title is not the proper value

File: web_src/js/features/repo-release.js (hunk 1 above)
Comment (lunny) [reply web_src/js/features/repo-release.js]: `input` will not be triggered but `change` will when pasting text on the input box.

File: templates/repo/release_tag_header.tmpl (hunk 2 above)
Comment (lunny) [reply templates/repo/release_tag_header.tmpl]: 04c9f07e73

File: web_src/js/features/repo-release.js (hunk 1 above)
Comment (wxiaoguang) [reply web_src/js/features/repo-release.js]: Are you sure "input" won't be triggered?

File: tests/integration/links_test.go [hunk 3]
```diff
@@ -36,6 +36,7 @@ func TestLinksNoLogin(t *testing.T) {
 		"/user2/repo1/",
 		"/user2/repo1/projects",
 		"/user2/repo1/projects/1",
+		"/user2/repo1/releases/tag/delete-tag",
```
Comment (wxiaoguang) [inline tests/integration/links_test.go]: Why use "delete-tag" to test? A "deleted tag" should not exist literally?

File: web_src/js/features/repo-release.js (hunk 1 above)
Comment (lunny) [reply web_src/js/features/repo-release.js]: e6ac5ed9a0, OK. Looks like `input` does also work.

Comment (wxiaoguang) [reply web_src/js/features/repo-release.js]: "input" and "change" are different.

File: tests/integration/links_test.go (hunk 3 above)
Comment (lunny) [reply tests/integration/links_test.go]: It's the only one existing record on release.yml which has `is_tag: true`.

Comment (wxiaoguang) [reply tests/integration/links_test.go]: Then I think it needs a comment, to say that is really exists

**Review by wxiaoguang — APPROVED** [review]
Not really LGTM (too many "if" tricks), but since there is a bug and the bug needs to be fixed.

File: tests/integration/links_test.go (hunk 3 above)
Comment (lunny) [reply tests/integration/links_test.go]: 03a3e9881c

**Pull request merged**